### PR TITLE
Add C++ ANTLR runtime

### PR DESCRIPTION
**Pull request by RokasUrbonasQC**
Adding C++ ANTLR code with bindings created using speedy-antlr-tool to significantly improve performance on large inputs.

Included changes:
 - Add C++ parsers autogenerated using the ANTLR tool and corresponding runtime
 - Integrate the C++ code using [speedy-antlr-tool](https://speedy-antlr-tool.readthedocs.io/en/latest/)
 - Make minor changes in grammar definition file:
   - Replace some rules using more advanced ANTLR features unsupported by `speedy-antlr-tool` with simpler equivalent rules
   - Fix some naming issues
 - Add a job for building package wheels to `build_and_publish` workflow
 - Update development documentation

Note for reviewers: All `.cpp` and `.h` files are autogenerated or directly taken from ATNLR repo.

### CODE REVIEW

**Comment by ivergara** [comment]
Is adding/vendor the ~300 files from the C++ parser really the only option?

**Comment by RokasUrbonasQC** [comment]
> Is adding/vendor the ~300 files from the C++ parser really the only option?

Yes, I think so. ANTLR Python target is slow, and our grammar definition is massive.

**Comment by ivergara** [comment]
> > Is adding/vendor the ~300 files from the C++ parser really the only option?
> 
> Yes, I think so. ANTLR Python target is slow, and our grammar definition is massive.

I'm not saying that using a C++-based parser is the bad part, just vendoring it into the repository. 

Would it be possible to use git submodules here? This would require that the C++ parser source code is also a git repository somewhere.

**Comment by RokasUrbonasQC** [comment]
> Would it be possible to use git submodules here? This would require that the C++ parser source code is also a git repository somewhere.

We could use a git submodule here. However, the C++ runtime is in the [huge ANTLR repo](https://github.com/antlr/antlr4), so I see three options here:
1. Add the whole ANTLR repo as a submodule. (It has a lot of files we don't need)
2. Write a script that would [sparsely checkout](https://git-scm.com/docs/git-sparse-checkout) just the files we need from ANTLR repo. (This would probably be slightly less convenient to use than a git submodule for anyone developing `pytsql`)
3. Keep the ANTLR runtime as plain files.

Let me know which option sounds the best, or suggest a different approach @ivergara.

**Comment by RokasUrbonasQC** [comment]
> > Would it be possible to use git submodules here? This would require that the C++ parser source code is also a git repository somewhere.
> 
> I could create a separate repo that would hold just the C++ runtime and parser files. Then we could use it as a submodule in `pytsql`. Does this approach sound good?
> 
> Alternatively, it might be possible to have all the C++ and `speedy-antlr-tool` code as a separate Python and Conda package, that would be a dependency of `pytsql`. (I might even be able to make the C++ version an "extra" for the package (e.g. `pip install pytsql[cpp-runtime]`)

After discussion with @ivergara via Slack, I have decided to leave the C++ files as they are right now, as other approaches are too difficult to implement or maintain, while providing little benefit.

File: environment.yml [hunk 1]
```diff
@@ -16,3 +16,4 @@ dependencies:
   - sqlalchemy
   - antlr-python-runtime==4.9.2
   - pyodbc
+  - mock
```
Comment (kklein) [inline environment.yml:19]: Do we need this even if we use `unittest.mock`?

Comment (RokasUrbonasQC) [reply environment.yml]: No, I might have added it by accident.

File: docs/source/development.rst [hunk 2]
```diff
@@ -63,12 +63,29 @@ Creating a grammar
 `antlr <https://www.antlr.org>`_, a parser generator. ``antlr`` expects grammar files
 and produces parsing Python code.
 
+Additionally, code generated using ``speedy-antlr-tool`` package is used to parse SQL scripts
+in C++ for better performance. The parsed tree is then converted into the Python equivalent.
+
 If you want to adapt the grammar please have a look at ``pytsql/src/pytsql/grammar/tsql.g4``.
-All other files in ``pytsql/src/pytsql/grammar`` are generated by ``antlr``. You can
-generate these files by running
+All files in ``pytsql/src/pytsql/grammar/cpp_src/antlr4-cpp-runtime`` are taken directly from
```
Comment (kklein) [inline docs/source/development.rst:70]: Have all the files from `github.com/antlr/antlr4/tree/master/runtime/Cpp/runtime/src` been copied or just a subset?
If a subset, how has this subset been determined?

Maybe a git permalink or a reference to a commit of the upstream code could be useful for the future.

File: docs/source/development.rst [hunk 3]
```diff
@@ -63,12 +63,29 @@ Creating a grammar
...
+the `ANTLR repository <https://github.com/antlr/antlr4/tree/master/runtime/Cpp/runtime/src>`_
+and the rest of the files in ``pytsql/src/pytsql/grammar`` are generated by ``antlr`` or
+``speedy-antlr-tool``. You can generate these files by running the following commands after
+adapting the path to your respective ``antlr`` jar file
```
Comment (kklein) [inline docs/source/development.rst:74]:
```suggestion
adapting the path to your respective ``antlr`` jar file as follows:
```

File: docs/source/development.rst [hunk 4]
```diff
@@ -63,12 +63,29 @@ Creating a grammar
...
+
+::
+
+    java -jar /usr/local/lib/antlr-4.9.2-complete.jar -Dlanguage=Cpp -visitor -no-listener -o cpp_src tsql.g4
+    java -jar /usr/local/lib/antlr-4.9.2-complete.jar -Dlanguage=Python3 -no-visitor -no-listener -o . tsql.g4
+
+And then running the ``speedy-antlr-tool`` using Python
```
Comment (kklein) [inline docs/source/development.rst:81]: From which directory should this be run?

File: setup.py [hunk 5]
```diff
@@ -17,9 +19,53 @@ def get_dev_timestamp(version: ScmVersion) -> str:
...
 
 
-setup(
-    use_scm_version={
-        "version_scheme": get_dev_timestamp,
-        "local_scheme": "dirty-tag",
-    },
-)
+def get_platform() -> str:
+    """Return the platform on which this code is run
```
Comment (kklein) [inline setup.py:23]:
```suggestion
    """Return the platform on which this code is run.
```

File: setup.py [hunk 6]
```diff
@@ -17,9 +19,53 @@ def get_dev_timestamp(version: ScmVersion) -> str:
...
+    """Create the ANTLR C++ extension to be passed to `setuptools.setup`"""
+
+    extra_compile_args = {
+        "windows": ["/DANTLR4CPP_STATIC", "/Zc:__cplusplus"],
+        "linux": ["-std=c++11"],
+        "darwin": ["-std=c++11"],
+        "cygwin": ["-std=c++11"],
+    }
+
+    sources = glob.glob("src/pytsql/grammar/cpp_src/**/*.cpp", recursive=True)
```
Comment (kklein) [inline setup.py:49]: Did you consciously decide against the `pathlib` module in favor of the `glob` module?

File: tests/unit/test_py_vs_cpp.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    assert len(py_context.children) == len(cpp_context.children)
+
+    # Validate children
+    for i in range(len(py_context.children)):
+        if isinstance(py_context.children[i], TerminalNodeImpl):
+            compare_terminal_node(py_context.children[i], cpp_context.children[i])
+        elif isinstance(py_context.children[i], ParserRuleContext):
+            validate_context(py_context.children[i], cpp_context.children[i])
+        else:
+            raise RuntimeError
```
Comment (kklein) [inline tests/unit/test_py_vs_cpp.py:86]:
```suggestion
            raise TypeError(f"Unexpected node type: {py_context.children[1]}")
```

File: tests/unit/test_py_vs_cpp.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            labels.append(m.group(1))
+
+    return labels
+
+
+def compare_common_token(py_token, cpp_token):
+    assert type(py_token) == type(cpp_token)
+    if py_token is None:
+        return
+    assert py_token.type == cpp_token.type
```
Comment (kklein) [inline tests/unit/test_py_vs_cpp.py:57]: Any reason for which you loop through keys/labels and access the data with `getattr` in line 91 and following and access them explicitly here?

**Review by kklein — COMMENTED** [review]
Thanks a bunch for your work - in particular for extending the documentation and writing tests for explicitly comparing our Python and C++ parsing!

Do you have thoughts on what ought to happen with the Python parser? Are you aware of cases in which in might still prove useful?

File: docs/source/development.rst (hunk 2 above)
Comment (RokasUrbonasQC) [reply docs/source/development.rst:70]: It was all files. I can add a link to the correct GitHub release.

File: setup.py (hunk 6 above)
Comment (RokasUrbonasQC) [reply setup.py:49]: Yes, using `glob` just resulted in less code.

File: docs/source/development.rst (hunk 2 above)
Comment (RokasUrbonasQC) [reply docs/source/development.rst:70]: Done

File: docs/source/development.rst (hunk 4 above)
Comment (RokasUrbonasQC) [reply docs/source/development.rst]: Updated the docs to include the answer to this

File: tests/unit/test_py_vs_cpp.py (hunk 8 above)
Comment (RokasUrbonasQC) [reply tests/unit/test_py_vs_cpp.py:58]: Added typing to the tests to hopefully make them easier to follow. `type` and other attributes accessed here are present in all `CommonToken` token instances, that's why they are accessed directly.

The labels accessed in line 91 can be anything that's defined in our grammar (any subclass of `ParserRuleContext`), so they are accessed dynamically using `getattr`.

**Comment by RokasUrbonasQC** [comment]
> Thanks a bunch for your work - in particular for extending the documentation and writing tests for explicitly comparing our Python and C++ parsing!
> 
> Do you have thoughts on what ought to happen with the Python parser? Are you aware of cases in which in might still prove useful?

The Python parser code just has to stay. It's still used in converting the C++ parse tree to the Python equivalent. I don't think there will be any other use for it.

I had an idea of using the Python parser code as a fallback in case no binary distribution exists for some specific OS + Python version combination. However, such cases are unlikely to occur with proper `cibuildwheels` and `conda` distribution workflows. I discussed this fallback idea briefly with @ivergara, and we decided that it would be too much effort to maintain.

File: tests/unit/test_py_vs_cpp.py (hunk 8 above)
Comment (kklein) [reply tests/unit/test_py_vs_cpp.py:58]: Thanks for elaborating!

File: docs/source/development.rst [hunk 9]
```diff
@@ -63,12 +63,33 @@ Creating a grammar
...
-All other files in ``pytsql/src/pytsql/grammar`` are generated by ``antlr``. You can
-generate these files by running
+All files in ``pytsql/src/pytsql/grammar/cpp_src/antlr4-cpp-runtime`` are taken directly from
+the `ANTLR repository release 4.9.2
+<https://github.com/antlr/antlr4/tree/5e5b6d35b4183fd330102c40947b95c4b5c6abb5/runtime/Cpp/runtime/src>`_
+and the rest of the files in ``pytsql/src/pytsql/grammar`` are generated by ``antlr`` or
+``speedy-antlr-tool``.
+
+You can generate these files by running the following commands in
+``pytsql/src/pytsql/grammar`` directory after adapting the path to your
```
Comment (kklein) [inline docs/source/development.rst:77]: Thanks :)

**Review by kklein — APPROVED** [review]
Thanks a lot @RokasUrbonasQC ! I think you did a great job in tackling this topic.

File: tests/unit/test_py_vs_cpp.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+FROM otherdb.dbo.othertable WITH(NOLOCK)
+WHERE kind = 1
+GROUP BY nr, short_nr;"""
+
+
+def get_rule_labels(context_cls):
+    init_func = context_cls.__init__
+
+    # Detect any context/token labels from the init function's assignments
+    # Conveniently, the type name is provided as a comment
```
Comment (EnchoMishinevQC) [inline tests/unit/test_py_vs_cpp.py:43]: Do we use the type comments in any way? I'd remove the comment if not.
(I know this is copied from the original validator, but I'm in favor of having the least amount of comments in code)

**Review by EnchoMishinevQC — APPROVED** [review]
Looks good to me! Thank you very much for working on this project!

File: tests/unit/test_py_vs_cpp.py (hunk 10 above)
Comment (RokasUrbonasQC) [reply tests/unit/test_py_vs_cpp.py:43]: It seems we don't use the type names provided in the comments. I'll remove this comment.
```suggestion
```

**Pull request merged**